Refactor js messages component to observables

mrtuvn opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

giving messages it's own ability array observable track by K.O and getting the data from the `customerData` messages
main purpose to Isolates from customerData (Default magento store messages in this object). Manage message in messages observables property only

Current magento get messages via `$.cookieStorage.get('mage-messages')` instead of internal `messages`. The first `if` block
default working for show messages. Logic here check messages from `cookieMessages`

![Screenshot from 2021-03-26 09-12-53](https://user-images.githubusercontent.com/1908873/112567541-97f6f880-8e13-11eb-9375-c61545466aad.png)
The second `if` block may never have a chance to do some render message. the variables messages may work but in another case it's may be cleared and not display message

the messages are retrieved from customerData. And next line is removing the customerData messages. This will also remove the values from the variable this.messages which is later on called in template.
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Theme/view/frontend/web/js/view/messages.js#L39

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#28692
Seem has closed
### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. ...
2. ...

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->
CC: @bassplayer7 maybe you can give me some reviews
@omiroshnichenko @nathanjosiah @zakdma
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

srsathish92 commented:
@magento run all tests

mrtuvn commented:
@magento run Unit Tests, Database Compare, Functional Tests CE, Functional Tests B2B, Functional Tests EE

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

omiroshnichenko commented on app/code/Magento/Theme/view/frontend/web/js/view/messages.js:
```diff
@@ -8,32 +8,48 @@
         defaults: {
             cookieMessages: [],
-            messages: [],
+            messages: ko.observableArray([]),
```
This is a backward incompatible change. If you want to make it "observable", use `tracks` instead. More information you could find [here](https://devdocs.magento.com/guides/v2.4/ui_comp_guide/concepts/ui_comp_uielement_concept.html)

mrtuvn replied:
I have added track variable in initObservable

mrtuvn replied:
Current when digging here i found that magento not pass message via this `messages`. It's empty may never used.
Magento get messages from `cookieStorage.get('mage-messages')`. Store error messages in local browser of customer

mrtuvn replied:
https://github.com/js-cookie/js-cookie
https://github.com/julien-maurel/js-storage
Maybe this one will be use in near future

What we current used https://github.com/julien-maurel/jQuery-Storage-API (seem no longer maintain). I'm not sure any workaround internal related this

gabrieldagama commented:
Hi @mrtuvn,

Can you provide a detailed explanation on how this issue can impact the customer? Please also add some steps to reproduce, the ones from the issue itself aren't very clear.

Thanks.

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Static Tests

mrtuvn commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, Static Tests

mrtuvn commented:
@magento run all tests

mrtuvn committed:
Refactor messages component to observables for update data


update


update js


update


update


up


update


remove whitespace


update


update


update

mrtuvn commented:
@magento run all tests

mrtuvn committed:
Update messages.js

Update code

mrtuvn commented:
@magento run all tests
